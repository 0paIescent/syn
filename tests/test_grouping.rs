#![cfg(all(feature = "extra-traits", feature = "full"))]
#![feature(rustc_private)]

#[macro_use]
extern crate syn;
use syn::{BinOp, Expr, ExprBinary, ExprGroup, ExprLit, Lit, LitKind};
use syn::token::Group;

extern crate proc_macro2;
use proc_macro2::*;

#[macro_use]
mod macros;

mod common;

fn tt(k: TokenNode) -> TokenTree {
    TokenTree {
        span: Span::default(),
        kind: k,
    }
}

fn expr<T: Into<Expr>>(t: T) -> Expr {
    t.into()
}

fn lit<T: Into<Literal>>(t: T) -> Expr {
    Expr::Lit(ExprLit {
        attrs: Vec::new(),
        lit: Lit {
            value: LitKind::Other(t.into()),
            span: Span::default(),
        },
    })
}

#[test]
fn test_grouping() {
    let raw: TokenStream = vec![
        tt(TokenNode::Literal(Literal::i32(1))),
        tt(TokenNode::Op('+', Spacing::Alone)),
        tt(TokenNode::Group(
            Delimiter::None,
            vec![
                tt(TokenNode::Literal(Literal::i32(2))),
                tt(TokenNode::Op('+', Spacing::Alone)),
                tt(TokenNode::Literal(Literal::i32(3))),
            ].into_iter()
                .collect(),
        )),
        tt(TokenNode::Op('*', Spacing::Alone)),
        tt(TokenNode::Literal(Literal::i32(4))),
    ].into_iter()
        .collect();

    assert_eq!(raw.to_string(), "1i32 +  2i32 + 3i32  * 4i32");

    assert_eq!(
        common::parse::syn::<Expr>(raw),
        expr(ExprBinary {
            attrs: Vec::new(),
            left: Box::new(lit(Literal::i32(1))),
            op: BinOp::Add(<Token![+]>::default()),
            right: Box::new(expr(ExprBinary {
                attrs: Vec::new(),
                left: Box::new(expr(ExprGroup {
                    attrs: Vec::new(),
                    group_token: Group::default(),
                    expr: Box::new(expr(ExprBinary {
                        attrs: Vec::new(),
                        left: Box::new(lit(Literal::i32(2))),
                        op: BinOp::Add(<Token![+]>::default()),
                        right: Box::new(lit(Literal::i32(3))),
                    })),
                })),
                op: BinOp::Mul(<Token![*]>::default()),
                right: Box::new(lit(Literal::i32(4))),
            })),
        })
    );
}

#[test]
fn test_invalid_grouping() {
    let raw: TokenStream = vec![
        tt(TokenNode::Literal(Literal::i32(1))),
        tt(TokenNode::Op('+', Spacing::Alone)),
        tt(TokenNode::Group(
            Delimiter::None,
            vec![
                tt(TokenNode::Literal(Literal::i32(2))),
                tt(TokenNode::Op('+', Spacing::Alone)),
            ].into_iter()
                .collect(),
        )),
        tt(TokenNode::Literal(Literal::i32(3))),
        tt(TokenNode::Op('*', Spacing::Alone)),
        tt(TokenNode::Literal(Literal::i32(4))),
    ].into_iter()
        .collect();

    assert_eq!(raw.to_string(), "1i32 +  2i32 +  3i32 * 4i32");

    assert_eq!(
        common::parse::syn::<Expr>(raw),
        expr(ExprBinary {
            attrs: Vec::new(),
            left: Box::new(expr(ExprBinary {
                attrs: Vec::new(),
                left: Box::new(lit(Literal::i32(1))),
                op: BinOp::Add(<Token![+]>::default()),
                right: Box::new(lit(Literal::i32(2))),
            })),
            op: BinOp::Add(<Token![+]>::default()),
            right: Box::new(expr(ExprBinary {
                attrs: Vec::new(),
                left: Box::new(lit(Literal::i32(3))),
                op: BinOp::Mul(<Token![*]>::default()),
                right: Box::new(lit(Literal::i32(4))),
            })),
        })
    );
}
