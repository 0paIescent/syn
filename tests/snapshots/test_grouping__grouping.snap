---
created: "2019-04-06T05:46:03.447914316Z"
creator: insta@0.7.4
source: tests/test_grouping.rs
expression: syntax_tree
---
Binary(
    ExprBinary {
        attrs: [],
        left: Lit(
            ExprLit {
                attrs: [],
                lit: Int(
                    LitInt {
                        token: Literal {
                            lit: 1i32,
                        },
                    },
                ),
            },
        ),
        op: Add(
            Add,
        ),
        right: Binary(
            ExprBinary {
                attrs: [],
                left: Group(
                    ExprGroup {
                        attrs: [],
                        group_token: Group,
                        expr: Binary(
                            ExprBinary {
                                attrs: [],
                                left: Lit(
                                    ExprLit {
                                        attrs: [],
                                        lit: Int(
                                            LitInt {
                                                token: Literal {
                                                    lit: 2i32,
                                                },
                                            },
                                        ),
                                    },
                                ),
                                op: Add(
                                    Add,
                                ),
                                right: Lit(
                                    ExprLit {
                                        attrs: [],
                                        lit: Int(
                                            LitInt {
                                                token: Literal {
                                                    lit: 3i32,
                                                },
                                            },
                                        ),
                                    },
                                ),
                            },
                        ),
                    },
                ),
                op: Mul(
                    Star,
                ),
                right: Lit(
                    ExprLit {
                        attrs: [],
                        lit: Int(
                            LitInt {
                                token: Literal {
                                    lit: 4i32,
                                },
                            },
                        ),
                    },
                ),
            },
        ),
    },
)
